File: projects/Dot-Dot_C#/Star.cs
using Microsoft.Xna.Framework.Graphics;
using Microsoft.Xna.Framework;
using System;
using System.Collections.Generic;

namespace Dot_Dot
{
    public class Star
    {

        private Texture2D sprite;   //create a texture for the sprite
        public Vector2 Position { get; private set; }  //create a vector for the position, public getter, private setter
        private Rectangle starBounds; //declare a rectangle to use for block collisions
        private bool noPreviousCollision; // declare a boolean to store if there was a previous collision 

        public Star(Texture2D starSprite, int x, int y)
        {
            // =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
            // Method				:	Star
            //
            // Method parameters	:	starSprite, x, y
            //
            // Method return		:	void
            //
            // Synopsis				:   This method is a constructor for the star object. 
            //
            // References			:   Challacade (2021) Make Games with MonoGame - Installation and Development Fundamentals. 
            //                          https://www.youtube.com/watch?v=sPH-sNTSrhw&ab_channel=Challacade
            // Modifications		:
            //							Date			Developer				Notes
            //							----			---------				-----
            //							2023-12-04		V. Arias				Star (constructor)
            //
            // =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
            sprite = starSprite; // declare the star sprite
            Position = new Vector2(x, y); // initialize the position vector
            starBounds = new Rectangle((int)Position.X, (int)Position.Y, sprite.Width, sprite.Height); // initialize the rectangle for collisions
            noPreviousCollision = true; // set the no previous collisions variable to false
        }

        public void Update(Dot dot, Texture2D starSprite, Texture2D starHitSprite, float objectSpeed, List<short> obstaclesXPositions, short[] lanes)
        {
            // =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
            // Method				:	Update
            //
            // Method parameters	:	obstaclesXPositions, dot, starSprite, starHitSprite, objectSpeed, lanes
            //
            // Method return		:	void
            //
            // Synopsis				:   This method updates the star object.
            //
            // References			:   Challacade (2021) Make Games with MonoGame - Installation and Development Fundamentals. 
            //                          https://www.youtube.com/watch?v=sPH-sNTSrhw&ab_channel=Challacade
            // Modifications		:
            //							Date			Developer				Notes
            //							----			---------				-----
            //							2023-12-04		V. Arias				Update
            //
            // =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
            Random random = new Random(); //declare and initilize the random object 
            bool noNewPosition = true; // set the neNewPosition boolean to true
            int newXPos = 0; // set the newXPos variable to zero
            short newLane; //declare a new lane short to get a random new lane for the star
            byte count; //declare a counter variable for looping

            Position = new Vector2(Position.X - objectSpeed, Position.Y); //update the objects position
            starBounds = new Rectangle((int)Position.X, (int)Position.Y, sprite.Width, sprite.Height); //update rectangle

            if (Position.X < -40) // completely off-screen
            {
                noPreviousCollision = true; //set noPreviousCollison to true
                newLane = lanes[random.Next(0, 2)]; //determine a new random lane for the star to be in
                while (noNewPosition)
                {
                    newXPos = random.Next(950, 1350); //determine a new random x position for the obstacle
                    for (count = 0; count < obstaclesXPositions.Count; count++)
                    {
                        if (!(newXPos >= obstaclesXPositions[count] - 60 && newXPos <= obstaclesXPositions[count] + 60))
                        {
                            noNewPosition = false; //if the new position doesn't overlap, set the noNewPosition variable to false
                        }
                    }
                }

                Position = new Vector2(newXPos, newLane); // Select a new random position for an obstacle/change to make them not overlap
                obstaclesXPositions[3] = (short)newXPos; // Store the new x position
                sprite = starSprite; //set the sprite equal to the star sprite
            }
           
            if (dot.DotBounds.Intersects(starBounds) && noPreviousCollision)
            {
                noPreviousCollision = false; //set noPreviousCollison to false
                sprite = starHitSprite; //set the sprite equal to the star hit sprite
                dot.score += 1; // If the dot collides with a star, the player earns a point
            }

        }

        public void Draw(SpriteBatch spriteBatch)
        {
            // =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
            // Method				:	Draw
            //
            // Method parameters	:	spriteBatch
            //
            // Method return		:	void
            //
            // Synopsis				:   This method draws the star object.
            //
            // References			:   Challacade (2021) Make Games with MonoGame - Installation and Development Fundamentals. 
            //                          https://www.youtube.com/watch?v=sPH-sNTSrhw&ab_channel=Challacade
            // Modifications		:
            //							Date			Developer				Notes
            //							----			---------				-----
            //							2023-12-04		V. Arias				Draw
            //
            // =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
            spriteBatch.Draw(sprite, Position, Color.White);                // draw the star sprite in the corresponding position
        }
    }
}

File: projects/Dot-Dot_Python/Star.py
import pygame #import the required libraries
import random

class Star (pygame.sprite.Sprite):
    def __init__(self, x, y):
        '''=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
        Method               :    __init__
    
        Method parameters    :    self, x, y
    
        Method return        :    none
    
        Synopsis             :    This method is a constructor for the Star object.
    
        References           :    Pygame (2023) Pygame Documentation. https://www.pygame.org/docs/
    
        Modifications        :
                              Date            Developer                Notes
                              ----            ---------                -----
                             2023-12-04        V. Arias                Initialization
    
        =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-='''
        super().__init__()
        try:
            self.image = pygame.image.load("assets/bigStar.png") # set the image to the star sprite
        except pygame.error as e:
            print(f"Error loading star image: {e}") #print a message if there was an error loading the image
        self.rect = self.image.get_rect() #define a rect for the star sprite
        self.rect.topleft = (x, y) #define the rect's position for the star sprite
        self.noPreviousCollision = True #set the noPreviousCllision variable to true

    def update(self, game, object_speed):
        '''=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
        Method               :    update
    
        Method parameters    :    self, game, object_speed
    
        Method return        :    none
    
        Synopsis             :    This method updates the star objects.
    
        References           :    Pygame (2023) Pygame Documentation. https://www.pygame.org/docs/
    
        Modifications        :
                              Date            Developer                Notes
                              ----            ---------                -----
                             2023-12-04        V. Arias                Update
    
        =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-='''  
        no_new_position = True #set the noNewPosition variable to true

        self.rect.move_ip(-(object_speed + 4), 0) #update the rect's position
        if  self.rect.x < -40: #star is completely off-screen
            self.noPreviousCollision = True #set no_previous_collision to true
            new_lane = game.lanes[random.randint(0,2)]
            while no_new_position:
                new_x_pos = random.randint(950, 1250) #determine a new random position for the obstacle
                for x_pos in game.obstacles_x_positions:
                    if not (new_x_pos in range(x_pos - 40, x_pos + 60)):
                        no_new_position = False #set no_new_positon to false
                        
            self.rect.topleft = (new_x_pos, new_lane) #update the rect with it's new random position
            game.obstacles_x_positions[3] = new_x_pos #store the new x position, offset by 3 to take cars into account
            self.image = pygame.image.load("assets/bigStar.png") #set the star's sprite to the big star
        if game.dot.rect.colliderect(self.rect) and self.noPreviousCollision: #using images, pygame.getrect()
            self.noPreviousCollision = False #set no_previous_collision to false
            self.image = pygame.image.load("assets/starHit.png") #set the star's sprite to the hit star
            game.dot.score += 1;    #if it collides with a star, they get a point
